src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs

```diff
@@ -275,6 +215,10 @@ internal XmlFormatClassReaderDelegate XmlFormatReaderDelegate
                     {
                         if (_helper.XmlFormatReaderDelegate == null)
                         {
+                            if (IsReadOnlyContract)
+                            {
+                                ThrowInvalidDataContractException(DeserializationExceptionMessage, null /*type*/);
+                            }
```

Comment:
I presume you aren't passing the UnderlyingType because this is only a problem on deserialization so don't want to remove the type handle from the typeToIDCache as it's still valid to be populated for the serialization case?